torch/nn/utils/parametrizations.py

```diff
@@ -285,6 +285,84 @@ def orthogonal(module: Module,
     return module
 
 
+class _WeightNorm(Module):
+    def __init__(
+        self,
+        dim: Optional[int] = 0,
+    ) -> None:
+        super().__init__()
+        if dim is None:
+            dim = -1
+        self.dim = dim
+
+    def forward(self, weight_g, weight_v):
+        return torch._weight_norm(weight_v, weight_g, self.dim)
+
+    def right_inverse(self, weight):
+        weight_g = torch.norm_except_dim(weight, 2, self.dim)
+        weight_v = weight
+
+        return weight_g, weight_v
+
+
+def weight_norm(module: Module, name: str = 'weight', dim: int = 0):
+    r"""Applies weight normalization to a parameter in the given module.
+
+    .. math::
+         \mathbf{w} = g \dfrac{\mathbf{v}}{\|\mathbf{v}\|}
+
+    Weight normalization is a reparameterization that decouples the magnitude
+    of a weight tensor from its direction. This replaces the parameter specified
+    by :attr:`name` with two parameters: one specifying the magnitude
+    and one specifying the direction.
+
+    By default, with ``dim=0``, the norm is computed independently per output
+    channel/plane. To compute a norm over the entire weight tensor, use
```

Comment:
I know this was here before but dim=None -> dim=-1 which computes the norm over the last dimension no?

Reply:
No, I believe weight_norm is special cased for this weird behavior lol.

```
>>> v = torch.randn(2, 2)
>>> g = torch.randn(())
>>> torch._weight_norm(v, g, 0)
tensor([[ 0.0577, -0.0345],
        [-0.0000,  0.0000]])
>>> torch._weight_norm(v, g, 1)
tensor([[0.0589, 0.0000],
        [0.0325, 0.0000]])
>>> torch._weight_norm(v, g, -1)
tensor([[ 0.0489, -0.0293],
        [ 0.0270, -0.0233]])
```

Reply:
May also be a good moment to eliminate older implicit behavior? and maybe simply require passing explicit dim?